src/wp-includes/option.php

```diff
@@ -827,7 +822,17 @@ function update_option( $option, $value, $autoload = null ) {
 	);
 
 	if ( null !== $autoload ) {
-		$update_args['autoload'] = ( 'no' === $autoload || false === $autoload ) ? 'no' : 'yes';
+		$autoload                = determine_option_autoload_value( $option, $serialized_value, $autoload );
+		$update_args['autoload'] = $autoload;
+	} else {
+		$raw_autoload = $wpdb->get_var( $wpdb->prepare( "SELECT autoload FROM $wpdb->options WHERE option_name = %s LIMIT 1", $option ) );
+		$allow_values = array( 'default-yes', 'default-no' );
+		if ( in_array( $raw_autoload, $allow_values, true ) ) {
+			$autoload = determine_option_autoload_value( $option, $serialized_value, $autoload );
+			if ( $autoload !== $raw_autoload ) {
+				$update_args['autoload'] = $autoload;
+			}
+		}
```

Comment:
Related to my reply in https://github.com/WordPress/wordpress-develop/pull/5671#discussion_r1419694609, we need to differentiate between "default/auto" vs "manual" behavior:
* If no autoload value is provided, WordPress chooses the default value automatically.
* If an autoload value is provided, WordPress shouldn't override it. In other words, the developer should have final say - otherwise, the entire API could be deprecated.

Therefore, this code checks whether the current value is a default/auto value, or an explicitly provided value, in which case WordPress shouldn't just alter it, as the developer explicitly set it.